### ISSUE TITLE
Collaboration (multi user) support

### ISSUE THREAD

**Issue by xPaw** [cdr/code-server#33]
It's not documented whether multiple users can use one self hosted installation at once or not. Is this a supported use case?

**Comment by sr229** [comment]
If I remember, this is a planned feature. But if this was a 1:1 code-oss port. VS Live Share could work on this.

**Comment by kylecarbs** [comment]
Currently, we do not support multiple users collaboratively editing on the same `code-server` instance. We've thought of this and would love to hear what your thoughts are. 

In our world, it'd be great to provide collaborative access, but never allow the collaborator to have control over the host machine in any way (for security reasons).

**Comment by sr229** [comment]
I think we should do this the way how VS Live share does it - RBAC by limiting views to read-only and restricting terminal access.

**Comment by xPaw** [comment]
I currently use https://github.com/c9/core for collaboration, where multiple users can log into the same server and work together.

code-server interests me as it uses VS Code so it heavily supported, unlike Cloud9. For my use case, permissions aren't really a problem.

**Comment by nhooyr** [comment]
@kylecarbs does vscode live share work?

**Comment by kylecarbs** [comment]
I believe it _would_ if the extension was open-sourced. Unfortunately, we can't provide support for it.

See: https://github.com/Microsoft/vscode/issues/38733#issuecomment-388213016

**Comment by zk-g** [comment]
This is a must for a cloud ide, Without this feature I don't even see an advantage of using a cloud editor as there are few ways to share the workspace through the client: and in plus require less resources than the browser.
So definitely +1

**Comment by zk-g** [comment]
Maybe also with an interactive button in the status bar showing connected user and options to manage permissions.
This could be a really game changer

**Comment by PhilsLab** [comment]
I'd have a look at how MS Word Online handles it:
![Word Online](https://user-images.githubusercontent.com/15452526/53984472-aaf9db80-4119-11e9-9616-7e88776a309c.png)

I like the idea of having a different cursor for every person actively editing,
this could be implemented via WebSockets / WebRTC.

Also, when providing a login functionality, there should be a modern backend for this, not just local users.
A solution that would fit very many people would be to integrate SAML and/or OpenID authentication.

As I mentioned in #85, you could use Keycloak (which can handle both SAML and OpenID) as an IdP:
Keycloak: [keycloak/keycloak](https://github.com/keycloak/keycloak)
Keycloak NodeJS integration: [keycloak/keycloak-nodejs-connect](https://github.com/keycloak/keycloak-nodejs-connect)

**Comment by baryla** [comment]
This feature is a must for us. My suggestions based on Cloud9 experience:

- Allow for a shared terminal

The way Cloud9 tackled this is great and it is something I'd personally love to see implemented in code-server.  

The collaboration feature is super important for us. We use Cloud9 as our go-to IDE but after Amazon's take over, we're looking for alternatives.

**Comment by kylecarbs** [comment]
How have they implemented it? Optimally it wouldn’t allow for full-access
(to prevent malicious activity).

**Comment by baryla** [comment]
This feature is a must for us. In terms of access, I would say that allowing the use of a terminal and editing any file/folder etc. would be enough.

My suggestions based on Cloud9 experience:

- Allow for a shared terminal
- See exactly where others are on a file (cursor, line highlight)
- A way to see how many people are currently viewing the project
- A way to see a person's active file (we usually talk about a particular file that may be nested in 10 folders) 

The way Cloud9 tackled this is great and it is something I'd personally love to see implemented in code-server.  We use it as our go-to IDE but after Amazon's take over, we're looking for alternatives and Coder/code-server would be it if it had collaboration/live share.

**Comment by baryla** [comment]
@kylecarbs - sorry I accidentally commented and then deleted it haha. See my new comment. 
Regarding how Cloud9 implemented this - I'm not sure about the self-hosted version as I haven't used however, the way we use cloud one is this:

1. We create a remote SSH workspace from our master account per project
2. From the master account, we invite all devs to have access to the workspace (files/folders + terminal)
3. Then you switch to your personal account and use the workspace shared by the master account. 

I'm trying to think what malicious activity there could be if this is self-hosted? Surely if you have a secure server and you trust your devs, there is nothing to worry about? In our case, we only have 3 full-time devs so I guess the only time where there could be a concern for this is if we get a freelancer or something. 

If you could share some of your concerns about this, I'd love to hear them! :)

**Comment by PhilsLab** [comment]
I think project selection should be done after login via a seperate page, locking each vs code session to one project to work on at a time.

**Comment by PhilsLab** [comment]
@baryla You could accidentially kill the server/container via the terminal, e.g. by installing packages that mess with the system, or killing the wrong process ID.
I already had some thoughts about that in #86.

**Comment by baryla** [comment]
@PhilsLab - 100% agreed about the project selection. When it comes to accidentally killing the server - I guess there could be times when this happens but I don't think this is a massive concern. Well, it would be if you worked directly from a live server. Ideally you should only be working on a dev or staging server and then only push your code live so it's not a big deal (IMO).

**Comment by PhilsLab** [comment]
My concern is not one of trust, it's one of software security. For an optimal experience, i'd really like tool see each terminal in it's own docker container, where you can select any docker image before running it. That would solve both security and the build/dev environment, and also is able to scale properly.
I'd use this as an instance on kubernetes to provide it to friends for example, where just building on trusting them not to screw up my server simply isn't enough :)

**Comment by nadavpa** [comment]
See how the jupyter team solved this: https://jupyter.org/hub

**Comment by babakness** [comment]
I believe codesandbox io has this feature as well and it is open source.

https://github.com/CompuIves/codesandbox-client

**Comment by sr229** [comment]
Be reminded we can try VS Live Share, which is available in the official extensions gallery, I think we can pull from there and support that atm

**Comment by NGTmeaty** [comment]
I don't think it'll be "officially" supported because I don't believe there's a official way to obtain it. (But I do think it's a very good solution if you can get the .VSIX.

**Comment by sr229** [comment]
[You can apparently](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/MS-vsliveshare/vsextensions/vsliveshare/0.3.1284/vspackage)

**Comment by baryla** [comment]
@sr229 - have you tried it? does it actually work?

**Comment by PassTheMayo** [comment]
@baryla 
![image](https://user-images.githubusercontent.com/16949253/54408888-e2d0c800-46b1-11e9-8373-8a85ea4be70c.png)

**Comment by baryla** [comment]
@PassTheMayo Great! haha. Thanks for letting me know that it doesn't work ;)

**Comment by zk-g** [comment]
The user interface may look like this:

![Code-server1](https://user-images.githubusercontent.com/17980844/54824618-c03f4000-4cab-11e9-821f-3dc48fb3d064.jpg)
![Code-server2](https://user-images.githubusercontent.com/17980844/54824630-c9301180-4cab-11e9-9db1-890b11d95edb.jpg)
![Code-server3](https://user-images.githubusercontent.com/17980844/54824639-cd5c2f00-4cab-11e9-8128-ca257295bac3.jpg)

**Comment by stepri** [comment]
I've installed VS Live Share via the following Dockerfile commands: 

```RUN mkdir -p /root/.code-server/extensions
RUN curl -JL https://marketplace.visualstudio.com/_apis/public/gallery/publishers/MS-vsliveshare/vsextensions/vsliveshare/0.3.1326/vspackage | bsdtar -xvf - extension
RUN mv extension /root/.code-server/extensions/ms-vsliveshare.vsliveshare-0.3.1326```

But I got the following error when starting a session:

```
Error: command 'liveshare.joinFromActivityBar' not found
```

**Comment by andreimc** [comment]
So I got vscode liveshare to work but there are a few caveats, you need to install normal vscode and x11 and vnc and you install the extension via vscode and you open vscode once, you symlink the extensions. It then works like supposed to via code server. I tried a few things without vnc like installing all the required packages it doesn’t seem to work. I have an ansible repo that sets up code server in this way: https://github.com/devonlineco/ansible-roles

**Comment by AnthonyMastrean** [comment]
I came here looking for information on multiple users sharing one code-server server, not for collaboration, but for each user to work on their own projects with proper isolation... I don't want to run a code-server per user with random ports.

Is there a feature/issue for this use-case?

**Comment by sr229** [comment]
This falls under collaboration feature as well so you already came to the right issue. However, for the isolation part - I don't think this is a possibility considering we only support use cases to run docker containers interactively @AnthonyMastrean

**Comment by AnthonyMastrean** [comment]
@sr229 All I need to do to make the Docker container into a service is add `-d` and a restart policy! But, I understand what you're saying. Is there anything under the Enterprise offering that might address this use-case?

**Comment by abubakar-bilal** [comment]
@andreimc I just cloned your ansible repo and tried to run this command

`ansible-playbook roles/code-server/tasks/main.yml `

But got this error
```
ERROR! 'unarchive' is not a valid attribute for a Play

The error appears to have been in '/root/ansible-roles/roles/code-server/tasks/main.yml': line 3, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: install code-server to lib dir
  ^ here
```

Any help on this would be helpful. How can I setup this on my server so that I can run code-server with live share extension.

**Comment by zk-g** [comment]
Is the multi-users function planned for any time soon?

Thanks

**Comment by sr229** [comment]
@zk-g yes its still planned to be done, but we're doing a rewrite of `code-server` using VSCode Web. That way we have a environment close to VSO (Microsoft's closed beta Online IDE service) which basically is just standard VSCode Web with Live Share.


I think that's a better route to go to rather than making our own. I wouldn't mind people reverse-engineering VS Live Share to make a Open Source implementation, it'll help us alot.

**Comment by Mustafa-C** [comment]
Since c9 appears to no longer be worked on and it's collaboration is shady many are looking for other web IDEs that support collaboration, it's one of the greatest advantages of having a cloud IDE. Hopefully this features is added in the months to come.

**Comment by sr229** [comment]
> Since c9 appears to no longer be worked on and it's collaboration is shady many are looking for other web IDEs that support collaboration, it's one of the greatest advantages of having a cloud IDE. Hopefully this features is added in the months to come.

As a matter of fact, I think I found a livecollab extension in the extension gallery that is open source, might worth checking?

**Comment by sr229** [comment]
Also our friends from the Theia project is working on a open source VSCode-compatible Livecollab extension, should be promising since I'm also pushing `code-server` for inclusion in Che.

**Comment by Mustafa-C** [comment]
Sounds interesting. What's the name of the extension you found?

**Comment by zk-g** [comment]
Can someone @notify me in this issue if they implement collaboration?
Thanks

**Comment by brandonkal** [comment]
@sr229 Also interested in the name of that livecollab extension. Thanks for looking into this.

**Comment by vanducng** [comment]
This feature will bring code-server to a new level, i just installed for some people in my team. It works like charm.

**Comment by brandonkal** [comment]
@vanducng what extension did you just install for this?

**Comment by jdabs** [comment]
Isn't this what git should be handling? I don't see why I'd want two developers working on the same file without seperate branches. Maybe in a learning environment, but not in a business setting.  

The git checkout feature should feed to the editor, only. If collaboration is needed, outside of an educational setting, then what for?

**Comment by AnthonyMastrean** [comment]
@jdabs when pair programming (or other similar practices)... and definitely in a business env. My team is full time remote. We screen share using a number of tools. VS live share has helped, but we’d love to use server based VS. 

https://www.jamesshore.com/Agile-Book/pair_programming.html

**Comment by sr229** [comment]
@jdabs I disagree with pair programming on the non-business side. Some companies that adopt Agile encourages this, Pair programming is quite useful in a practical sense:

- Situation 1: you need live assistance with the person in charge of that subsystem, and you're not present in the office or the engineer in charge isn't.

- Situation 2: when crunching, having a dev environment already setup and people to connect to is easier, especially if its a minute change and the codebase is too large to pull, especially in constrained bandwidth environments (4G, EDGE, etc.)

**Comment by nhooyr** [comment]
This is unlikely to occur at the moment. I'll be sure to open an issue for this when the time is right.

**Comment by MatthiasHertel** [comment]
@nhooyr whats up with this ...

i guess corona virus **"makes the time right"** for the strong demanded collab feature. it should be possible use collab on premise instead of using live share.

https://github.com/cdr/code-server/blob/master/doc/FAQ.md#collaboration
from the faq.md
> Collaboration
> At the moment we have no plans for multi user collaboration on code-server but we understand 
> there is strong demand and will work on it when the time is right.

**Comment by nhooyr** [comment]
Yes, you're absolutely right. Unfortunately at the moment the team is occupied but I am reopening this. Hopefully we can get to it within the next few months.

**Comment by UnlimitedCookies** [comment]
I don't know if this helps.
Potentially the javascript-based backend [shareDB](https://github.com/share/sharedb) might be a good fit for syncing the texts between the clients.

**Comment by GalacticLion7** [comment]
How about using [TogetherJS](https://github.com/jsfiddle/togetherjs) by Mozilla?

**Comment by nhooyr** [comment]
That looks really cool! Unfortunately the examples don't seem to work for me. I tried to use my invite link in a separate private browser tab and couldn't get the two to connect. If I can get it to work, it seems like it'd be a great low cost solution.

**Comment by mgielda** [comment]
What about yjs - https://github.com/yjs/yjs ? It has Monaco support: https://github.com/yjs/y-monaco

**Comment by umcconnell** [comment]
@nhooyr The TogetherJS examples aren't working, because the examples try to connect to `hub.togetherjs.com`. This hub doesn't exist anymore. [This issue](https://github.com/jsfiddle/togetherjs/issues/1172) gives more details on the problem and instructions on how to set up a hub.

**Comment by ribrdb** [comment]
It looks like firepad has support for Monaco, I wonder if that could work.
https://github.com/FirebaseExtended/firepad/blob/master/examples/monaco.html

> **What is Firepad?**
> Firepad is an open source real-time collaborative text editor. It provides true collaborative editing, complete with intelligent operational transform-based merging and conflict resolution.
> **How is Firepad different than other collaborative text editors?**
> Most collaborative text editors require special code to run on a server, making them impractical to use if you're not already using the right server stack. Firepad has no server dependencies and instead relies on the Firebase Realtime Database for real-time data synchronization. This means that it's easy to add Firepad to any application, even static websites. All you need to do is drop in the JavaScript files and go.

**Comment by OliverwengFiltered** [comment]
@kylecarbs @nhooyr 

Hi there,

wondering if you guys have any new updates regarding to this live share.

the terminal is not synced currently, so we have to do a screen share additionally when it comes to live collaboration.

it would be great if the code server itself has this feature supported.

we can discuss about a sponsorship if that's something you guys interested in.

Thanks
-Oliver

**Comment by rdbeach** [comment]
Hi. I have developed a real-time collaboration system for code server based on Firepad/Firebase.

You can see a demo here:

https://youtu.be/fW9oXbbHjC8

I plan on shopping this around a bit to see if there are any companies that are interested, and would like to work with me on integrating this feature into there business. 
rdbeach@gmail.com

**Comment by caio-vinicius** [comment]
+1 this really will help

**Comment by rdbeach** [comment]
Hi. the collaboration system using firebase is now open source. You can set up multiple accounts for the same code-server instance, each with their own settings! You can collaborate in real time with team members. There is a user tracking system so you can see where everybody is and what they are doing, and you can quickly focus in on their work.

I have released a beta version. Get it here: https://github.com/coderpair/vscode-live  
Looking for some people to try it and give me feedback.

<img width="986" alt="screenshot2" src="https://user-images.githubusercontent.com/8367977/100521947-0eccff00-315c-11eb-8948-d606815956bc.png">

Also looking for help in general as to long term maintenance and/or incorporating it into code-server. It's a big mod, almost doubling the size of the patch, so I don't know if it make sense to work it in or just continue to develop in parallel.

**Comment by nhooyr** [comment]
> Hi. the collaboration system using firebase is now open source. You can set up multiple accounts for the same code-server instance, each with their own settings! You can collaborate in real time with team members. There is a user tracking system so you can see where everybody is and what they are doing, and you can quickly focus in on their work.
> Also looking for help in general as to long term maintenance and/or incorporating it into code-server. It's a big mod, almost doubling the size of the patch, so I don't know if it make sense to work it in or just continue to develop in parallel.

This is great news! I'll try and take a look soon and we what we can do!

**Comment by code-asher** [comment]
This is amazing. Are there any technical limitations that would prevent
this from being a VS Code extension instead of patching? It would be
awesome to distribute it as a default built-in extension or something.

**Comment by zepinto** [comment]
Awesome, what kind of help do you need?

**Comment by rdbeach** [comment]
@code-asher  @zepinto Im not saying it can't be done as an extension, but I couldn't figure out how to do it. I doubt that it can be done, at least with a Firepad based solution. 

I originally tried to create it as an extension but but was frustrated by the lack of low level access. With a Firepad based solution, you need to have access to an ICodeEditor instance, and they don't even give you that in the extension API. I also ended up having to make a few fundamental changes to the way the ICodeEditorWidget works. One more thing: with the API, they don't let you interact with the DOM at all (probably with good reason!) so I tapped into some native code to give the user list more of the look and feel of vscode.

As for right now, I would like to know if the guys at coder would be open to bringing in some of the server side changes I have made? These changes are small--like adding options to the configuration for Firebase. Even the multi-user option is fairly straightforward and could easily be implemented right now. I also put back the dashboard that had previously been taken out. If you brought in these changes, then I would basically only have to keep the patch up to date, which I can easily do by myself. Then we can figure out what to do with the patch or how to make it more modular later.

One thing: I started working on this before you changed the whole thing to express, and I have yet to look at those changes.

Lastly. My original intent was to create a derivative work from vscode and incorporate it into some kind of web based service. I have an idea of what I want to do, but I'm kind of looking for someone to team up with. I actually talked with a few code-interview businesses about using this, but I didn't get any takers as of yet.

**Comment by bpmct** [comment]
@ribrdb I just gave this a shot and it works quite well! I noticed that different passwords will allow you to log in as different users, but there is no username field in the login page.

@code-asher @nhooyr can we support this (and future projects) as "plugins" so it doesn't have to be done as a patch? I understand that some work was done to support this, but I'm not up to date with the current state of this :)

**Comment by bpmct** [comment]
@rdbeach I just gave this a shot and it works quite well! I noticed that different passwords will allow you to log in as different users, but there is no username field in the login page.

@code-asher @nhooyr can we support this (and future projects) as "plugins" so it doesn't have to be done as a patch? I understand that some work was done to support this, but I'm not up to date with the current state of this :)

**Comment by nhooyr** [comment]
> @code-asher @nhooyr can we support this (and future projects) as "plugins" so it doesn't have to be done as a patch? I understand that some work was done to support this, but I'm not up to date with the current state of this :)

That work was unrelated. Here we're talking about VS Code plugins. In https://github.com/cdr/code-server/pull/2252 we added an API that allows "application plugins" that e.g. add Jetbrains IDE support to code-server in the enterprise product.

@code-asher and I will need to take a look at @rdbeach's code before we can say whether or not it can implemented as an VS Code extension or if we need to extend the API to allow it to be implemented as such.

**Comment by rdbeach** [comment]
This is what I am looking for in terms of a plugin API. It is particular to my needs, but I think it could be made broad enough to be generally applicable.

1) Allow custom plugin configuration options which would be accessible both on server and client side.

2) A UI for interacting with the plugin. This can be built by the plugin, but code-server should provide a place for it to live. Preferably a centralized place.

3) Plugins should be able to hook into some of code-server's functionality. My particular mod has an option to temporarily disable code-server. 

Will be bringing the mod up to date with the most recent version of code-server soon. Probably not this week though since I am pretty busy. 

Has anyone given consideration to multiple-user logins? Right now I am offering a very simple system that allows for multiple passwords. Each password logs a user in under a different account, and each account has their own settings folder. At the moment, there isn't even a way to log-out.

**Comment by KatiGithub** [comment]
Well, the [live-share](https://github.com/MicrosoftDocs/live-share) extension for Visual Studio Code is now open and code can be seen.

**Comment by daumling** [comment]
I've implemented nginx authorization via its auth module. It would be great if I could set a HTTP header to have code-server open a file, folder or workspace. I've opened feature request #2506 with more details.

**Comment by sr229** [comment]
If you're interested you can use the APIs Live Share uses which is in the code base already.

**Comment by gsmachado** [comment]
@sr229 @daumling 
Maybe it's totally unrelated to this subject, but I tried `MS-vsliveshare.vsliveshare` [extension](https://marketplace.visualstudio.com/items?itemName=MS-vsliveshare.vsliveshare) with `code-server` and I couldn't make it work because of some kind of authentication issue.

I described what happened [in this discussion topic](https://github.com/cdr/code-server/discussions/2530).

**Comment by prosenjit-manna** [comment]
I am able to create multiple instance with multiple local user account and setup in aws ubuntu server. Each user have their own directory permission.

**Comment by sr229** [comment]
> I am able to create multiple instance with multiple local user account and setup in aws ubuntu server. Each user have their own directory permission.

This is out of topic and not related to the actual feature request.

**Comment by pojntfx** [comment]
To mitigate this issue, I wrote [pojde](https://github.com/pojntfx/pojde), which creates multiple isolated code-server instances with Docker (or another CRI implementation) and handles updates, authentication, volumes, toolchains, TLS etc. - although proper multi-tenancy support (and especially real-time collaboration) would still be awesome :)

**Comment by kainzpat14** [comment]
This seems to be possible as an extension without any additional APIs, I have been experimenting with [Rijul5/vscode-teletype](https://github.com/Rijul5/vscode-teletype) and it seems that the decoration API and the editor API both provide everything one would need. 

The current state of the experiments can be found in [kainzpat14 /
vscode-teletype](https://github.com/kainzpat14/vscode-teletype). The original version was with Teletype, which is a bit to opinionated for my tastes and also uses proprietary APIs, so I have been trying to get YJS (which was mentioned in this thread) working instead, it seems to do what one would need. 

VSCode has a few unfortunate behaviours that make something like this a bit difficult, such as the fact that it seems to be multithreaded (which a node program shouldn't be) and provides no way to mark edits as "system" generated, leading to race conditions, for which I have yet to come up with a nice synchronization mechanism. 

All in all it appears to be very possible and I hope to continue working on the extension to provide both code-server internal co-editing and sharing between VSCode and code-server.

**Comment by jsjoeio** [comment]
> Liveblocks is real‑time collaboration infrastructure as a service

https://liveblocks.io/blog/introducing-liveblocks

I don't know if this could be integrated into code-server but worth posting here so I don't forget.

**Comment by pojntfx** [comment]
One solution I have gotten to work: https://www.codetogether.com/

Support pretty much everything live share supports, but also works in (just tested it):

- code-server
- VSCodium
- VSCode

Plus IntelliJ and Eclipse, although I haven't tested it. They also allow anonymous usage w/o an account which is pretty nice and they publish to Open-VSX, so it's installable from the "Extensions" tab: https://open-vsx.org/extension/genuitecllc/codetogether

I think the last part to make code-server support it 100% is having something like VSCode's `argv.json` to set the required proposed API flags (`-enabled-proposed-api`) and this would be pretty much a 1:1 replacement for live share :)

**Comment by jsjoeio** [comment]
> I think the last part to make code-server support it 100% is having something like VSCode's argv.json to set the required proposed API flags (-enabled-proposed-api) and this would be pretty much a 1:1 replacement for live share :)

@code-asher any ideas on how difficult this would be to implement?

**Comment by code-asher** [comment]
That flag could be set when spawning code-server or in the config.yaml.
Although perhaps I have misunderstood the requirement.

**Comment by KaKi87** [comment]
Here's recent activity occuring, then. Have a nice week !

**Comment by AndyIsHereBoi** [comment]
i would love to see this. it would be basicly replit but better.

**Comment by toddewilliams** [comment]
@AndyIsHereBoi There are already two solutions that provide this for code-server, ([Duckly](https://duckly.com/) and [CodeTogether](https://codetogether.com)). 

The proper way to set each of them up is documented on the [collaboration page](https://coder.com/docs/code-server/latest/collaboration). Each website has the detailed capabilities of each so you can see which on is a better choice for you.

Also, here is [a multi-product comparison chart](https://www.codetogether.com/docs/pair-programming-tools-for-remote-development/) that includes both of those products (as well as others), to help you decide what solution is best for you and your team.

**Comment by KaKi87** [comment]
> There are already two solutions that provide this for code-server

Don't these solutions require separate VSCode instances ?

**Comment by toddewilliams** [comment]
@KaKi87 I've never used Duckly but since CodeTogether is my company's product I can answer definitively for it.

You will not need a separate VS Code instance with CodeTogether. 

Once you've followed the installation procedures on the collaboration page, you can start hosting a CodeTogether session from the CodeTogether view and then share the URL with others. Anyone with the URL can join using either a browser or an IDE (VS Code, VSCodium, Eclipse, IntelliJ, or an IDE based on one of those). 

Here is a [quickstart guide](https://www.codetogether.com/docs/three-steps-to-pair-programming-vs-code/) for the process for VS Code. Just skip Step 1 (install) and use the [installation instructions ](https://coder.com/docs/code-server/latest/collaboration)from the code-server documentation instead.

**Comment by pkkid** [comment]
While both solutions mentioned above are a great stop-gap, I also wanted to chime in and say that they are also less than ideal for a few reasons. I'm adding my comments here in hopes that the idea of moving toward an implementation to collaborate within code-server doesn't die because of the comments above. I don't want this post to sound like a rant, but more a "please don't let this request be swept under the rug."

1. Both solutions mentioned require a third-party external account as well as limitations on how long or what we can share unless paying for an account.
2. The workflow of logging into another account, then sharing a link together is jarring and confusing when we just want to see the other user's cursor and potentially follow their location.
3. Both options are more like using a screen sharing tool than they are working together on the same code like Replit because of the host/client connection process.
4. Sharing with Duckly, I couldn't get working correctly from the same instance of code-server. It times out after accepting and waiting to connect.
5. Sharing with a CodeTogether link actually opens a second IDE on their website. So there are technically two IDEs, and code being send to third-party servers and none of the host IDE settings transfer over.

**Comment by toddewilliams** [comment]
@pkkid **I'm not going to comment on your feedback for the code-server team as that feedback is for them.**

I would like to simply clear up a couple of the concerns you expressed specifically about CodeTogether.

> Both solutions mentioned require a third-party external account as well as limitations on how long or what we can share unless paying for an account.

CodeTogether provides completely anonymous (no account) and free usage for an unlimited number of sessions for up an hour and four participants. Sessions that are longer or larger than that would require one of our subscription plans.

>The workflow of logging into another account, then sharing a link together is jarring and confusing when we just want to see the other user's cursor and potentially follow their location.

An account to join a CodeTogether session is never required. Any session at any level can always be joined with just a URL.

>Both options are more like using a screen sharing tool than they are working together on the same code like Replit because of the host/client connection process.

I don't know what you mean in CodeTogether's case because it does has the same "Google Docs-like" editing that Replit does, as well as chat and audio/video if you want it. You're literally editing the same files at the same time in the editor of your choice and no screen sharing is involved.

>Sharing with a CodeTogether link actually opens a second IDE on their website. So there are technically two IDEs

I believe there is confusion here. When you join a CodeTogether session, there is no "second IDE on their website". You have an IDE-like experience in your browser, or in your actual IDE. If you join from an IDE, then there are technically two IDEs, one for the host and one for the participant since they joined from it. 

>and code being send to third-party servers 

This is somewhat true in that the minimal amount of content is sent through the servers in an E2E-encrypted channel that they don't store and can't decrypt. There are the [details on the end-to-end security of CodeTogether ](https://www.codetogether.com/download/security/) for the interested.

>and none of the host IDE settings transfer over.

That's true and by design. Each participant chooses the theme and keybindings they prefer from their favorite IDE. The host has one set of preferences and each participant can choose something else.

If @pkkid or anyone else has any questions about [CodeTogether](https://codetogether.com), the development team answers questions on [Gitter](https://gitter.im/CodeTogether-com/community) and here's the [GitHub project](https://github.com/Genuitec/CodeTogether) for any bugs or enhancements.

**Comment by AndyIsHereBoi** [comment]
yea but it would be nice to have it on the app itself. either as a thing i can enable and then just share a link (like one of those other apps) and i can set a password for each one to use (and expiration). in my opinion i would like it to be on code server because it would just be easier for my use case. but that depends on what anyone else uses it for i guess

**Comment by GalacticLion7** [comment]
E

**Comment by KaKi87** [comment]
> you can start hosting a CodeTogether session from the CodeTogether view and then share the URL with others. Anyone with the URL can join using either a browser or an IDE

So, the collaborators won't use the host's VSCode instance, which is my point.

Therefore, those options are not *solutions* but *alternatives*.

The feature requested here implies that all collaborators directly interact with the self-hosted code-server instance without any third-party dependency.

**Comment by AuthorShin** [comment]
Any update on this?
Aside from collaboration, is there a way for different users to sign in with their own credentials to access their own workspace?

**Comment by jsjoeio** [comment]
@AuthorShin nothing besides what we have in [the docs](https://coder.com/docs/code-server/latest/collaboration)

**Comment by sylvaticus** [comment]
I follow the AuthorShin's question.. to be clear, in our lab server we have JupyterHub that manages the login and multi-user deployment of Jupyter for system users, and wondering if there is something similar for visual-studio code server...

**Comment by UlyssesZh** [comment]
+1 for this. Something like JupyterHub helps.

**Comment by Jerry080801** [comment]
What about vscode live

**Comment by toddewilliams** [comment]
@Jerry080801 Using VS Code Live Share is not legal. I cover exactly why that is [here](https://github.com/microsoft/vscode/issues/31168#issuecomment-1136375670).

**Comment by paccerdk** [comment]
@toddewilliams I don't want to derail the subject of this issue, but it's not clear to me from what you shared, that its not legal - it does states that "You may install and use any number of copies of the software to use solely with **Microsoft Visual Studio family of products**."
But is this not a Microsoft Visual Studio family product?

**Comment by toddewilliams** [comment]
> But is this not a Microsoft Visual Studio family product?

**Short answer:** No, as it's not a product from Microsoft.

**Longer answer:** Microsofts lists those products in their Terms of Use for their website, which I referenced in my prior link. And they also have this handy [page for documentation for the "Microsoft Visual Studio Product Family"](https://learn.microsoft.com/en-us/visualstudio/?view=vs-2022). You can see the members of the product family are both well-known and quite precisely restricted to Microsoft products.

**Comment by paccerdk** [comment]
It's pretty clear for me what the Visual Studio product family is, its just a little fuzzy for me when a product stops being Visual Studio Code (in this case, running in the browser), and starts being a "new" product that is no longer Visual Studio Code. 
I'm going to go ahead and assume that you know. Thanks for the feedback.

**Comment by danopia** [comment]
Further explanation can be found on the Code-Server FAQ:
https://coder.com/docs/code-server/latest/FAQ#why-cant-code-server-use-microsofts-extension-marketplace

**Comment by KaKi87** [comment]
Anyway, I tried it and it doesn't work, so we can forget it.

Still looking forward to `code-server` featuring collaboration.

**Comment by reyo** [comment]
> I currently use https://github.com/c9/core for collaboration, where multiple users can log into the same server and work together.
> 
> code-server interests me as it uses VS Code so it heavily supported, unlike Cloud9. For my use case, permissions aren't really a problem.

I'm in the exact same spot. Would really really like to see multi user support. Cloud9 is kinda dead, maybe, somewhat - underdeveloped and badly maintained under Amazon rule.

**Comment by matbgn** [comment]
This feature seemed to be getting a lot of traction and was really close to being solved by @rdbeach back in 2020: 

> I have released a beta version. Get it here: https://github.com/coderpair/vscode-live Looking for some people to try it and give me feedback.

My question is, is there any fundamental opposition from @toddewilliams or anyone else, or is it just a case of bad luck that this issue has been stalling for exactly 4 years? Yet @nhooyr seemed to show some interest at one point, if I followed correctly.

**Comment by cnicodeme** [comment]
I agree that it's surprising that this feature never saw the day of light!

@baryla, you mention this being a must for you ([here](https://github.com/coder/code-server/issues/33#issuecomment-470816448)). Can I ask what are you using nowadays that meet your needs? You mentioned Cloud9 on AWS but were looking to change. (We're on the same boat)

**Comment by toddewilliams** [comment]
@matbgn I'm responding since I was specifically mentioned. 

>My question is, is there any fundamental opposition from @toddewilliams or anyone else

Not only don't I have any opposition to this, I'm not even in a position to have any opposition. I don't contribute to code-server and am not an employee of Coder.

I think all approaches to this issue are great! We ([CodeTogether](https://codetogether.com)) developed our technology to enable cross-IDE collaboration between **all** the IDE's developers use (that are built on VS Code, Eclipse, or IntelliJ). As a result, it happens to work for code-server so I contributed to the documentation to make everyone aware that it was one solution to this issue.

My other posts simply point out that since code-server is not a Microsoft product, that it is prohibited from using Microsoft's Live Share as a solution due the the license under which Live Share is published.

So while I haven't looked at your project, I certainly applaud that you're trying to solve this issue as well and wish you nothing but success with it.

**Comment by code-asher** [comment]
On the Coder end our scope has shrunk to maintaining what is already there and keeping Code up to date so adding collaboration is likely not something we would drive at the moment.

That said, as far as I know there is no fundamental opposition to collaboration.  I do have some opposition to adding it directly to code-server though (rather than as an open-source extension) unless there is a compelling use-case that cannot be supported by extensions.

**Comment by KaKi87** [comment]
How could the authentication system be enhanced from an extension ?

**Comment by code-asher** [comment]
I am not too familiar with VS Live Share but we could do authentication the same way they do, which I think is based on sending a link with a token in it.  What sort of enhancements are you thinking about?

**Comment by KaKi87** [comment]
> we could do authentication the same way they do, which I think is based on sending a link with a token in it

Would that allow using VCS, plugins, etc. ?

> What sort of enhancements are you thinking about?

Well, as the title says : multi user.
Like operating systems, cloud platforms, VCS platforms, etc.

**Comment by danopia** [comment]
> > What sort of enhancements are you thinking about?
> 
> Well, as the title says : multi user.

I think Multiuser is a separate goal from collaboration. Maybe we can clarify what is requested..

Various collaborative editors - including VSCode live share - still have each user running their own local editor. They are simply connected together with one-off tokens and I imagine that could be achievable via plugin here too. Every user would be expected to have an editor (Code-Server instance) already.

Meanwhile, multiuser I understand to be actually multi-tenant where one server would run multiple user's editors at the same time. If I understand that correctly then it's a more foundational task? But if you don't care about live collaboration, you could just run Code-Server multiple times and give each user their own instance. And then this hypothetical plugin could still be added to collaborate, once it exists

**Comment by KaKi87** [comment]
Actually, I meant both : live collaboration using multiple users, which would be like Google Workspace but for code instead of office documents.

Users have access to the same UI on which they log in and open files.
Collaboration is just multiple people opening the same file at the same time.

What's so great about code-server is that everything happens server-side, opening this possibility.
Meanwhile, creating a plugin that just provides link-based secondary UIs for guest users would not give any advantage over abovementioned alternatives.

**Comment by reyo** [comment]
Good example of what people expect, I think, is like AWS Cloud9. We use it. It's like Google Docs but for code. Filesystem is shared between all users and I can see other colored cursors in my code while I edit. Even terminals are shared. Cloud9 base is opensource, or was, until AWS came ... they should be still available on Github. Why I'm looking at code-server and comment on this issue is that we'd like to break free from AWS :). They aren't that great company, but Cloud9 is still, the only software that does the thing that we all love in our company (true live single file collabration).

**Comment by KaKi87** [comment]
Then what would happen when multiple users would open the same file without using the plugin ?

**Comment by code-asher** [comment]
We could just do nothing and let it work the same as if there were two separate code-server instances running.  So whoever saved last would have the latest.

We could also make it so guest instances have no file access, then they have to use the plugin to edit forcing them to use it in collaboration mode.

**Comment by KaKi87** [comment]
What guest instances ?

**Comment by code-asher** [comment]
The way I imagine the plugin working is you would create a token to give to people (through the API we would add) and those would be "guest" instances.

**Comment by morganwk** [comment]
Is there really _no_ solution that doesn't rely on either Live Share, Duckly, or CodeTogether?

This feature as well as remaining FOSS and not tied to a paid service are both mission critical for me. 

Would it be at all feasible to patch in something like Etherpad, even if it isn't fully featured, just for syncing the state of the text buffer across open file(s)? Just... something? Anything?

I realize the security implications there but I'm just spitballing in the hopes that a band-aid solution might be somewhat attainable even if a native feature is realistically many months or years down the line.

**Comment by toddewilliams** [comment]
@morganwk Your comment makes me curious. 

OSS and free are two different concerns. CodeTogether is not OSS, because we made a large investment in both terms of technology innovation, engineering, and server time and want to be able to pay our developers and cover our worldwide hosting costs. We do this by offering a very low-priced "freemium" solution. Our [free version](https://www.codetogether.com/live/pricing/) includes unlimited 1-hour sessions with up to 3 people. Most of our users find that the free version completely meets their needs. But if it doesn't, you can remove those limits completely for a few dollars per month.

So free is already available, but not as OSS because we have expenses to pay. And even if CodeTogether was OSS, someone would still have to provide a worldwide server network to run it at scale, so the expenses would remain. 

Can you help me understand why OSS in this particular area (which you'll be using only to develop a solution, not deliver as part of a solution) is so "mission critical" for you? I really am simply trying to understand as you're not the only one who has this belief.

**Comment by morganwk** [comment]
In all honesty, it's largely because I'm broke. But there are a host of reasons. Pun intended.

The OSS aspect of things is less important - although I'll get to that later - but... the fact of the matter is that you are not in fact offering a _free_ solution. And that's not your fault.

Imposing a limitation 3 collaborators for one hour at a time is reasonable for the service you're providing, being that you're offering them in such a way that they incur expenses that must be recovered and turn a profit in order to self-sustain.

Unfortunately, those limitations simply do not meet my needs. 

Most of the time, I'll only be working with one other person at a time. But it isn't uncommon for me to need to open up a document with 5 or more people simultaneously. That capacity is non-negotiable.

And, in either case, I _rarely_ do so for only one hour at a time. Even if the sessions are unlimited, having to exit and restart them at an hourly interval represents an absolutely intolerable interruption to my workflow - maybe I'm just overly sensitive to that because of ADHD, but as I'll say several times throughout, it's the principle of the matter.

That brings me to the money side of things.

I'm trying to cut down on subscription-based products and services. The harsh reality is that in a world that's being absolutely ravaged by price gouging and inflation, $5.25 billed monthly may sound inconsequential, but it adds up VERY quickly when combined with every other "reasonably priced"* piecemeal solution we've all steadily come to expect and rely upon.

*_reasonably priced is in scare quotes here specifically because the price only becomes unreasonable when combined with many other subscriptions, even though it's far more than reasonable in isolation_

Of course, given your particular offerings and infrastructure, not only is this a reasonable price, in isolation, but I would be highly unreasonable to expect you to provide the same services for completely free because you'd be operating at a loss. What I'm looking for - even if it isn't quite as feature complete or well-integrated, is a local solution that doesn't rely on someone else's servers.

---

The OSS nature of such a solution is mostly in relation to it not being something that exists for profit - because even the absolute cream of the crop of free closed-source software is subject to ruination at a moment's notice as a direct result of the whims of capital.

There's no guarantee that CodeTogether will exist forever. If anything, there's a guarantee that it _won't_ simply because of the linear nature of time and the life cycle of private entities existing within the structure of capitalism - they either get bought up, get too big to fail (all things being relative), or they fail.

There also the exists the issue of OSS and extensibility - aside from the fact that you may change how your service works at any given moment without notice (assuming your TOS allows for that), there's also no guarantee I won't eventually face a situation where I need a functionality that is an uncommon use case and thus won't be considered for development, and I'll have no choice but to migrate.

That is highly unlikely, and it looks like your services are robust especially for the low price point, but it's the principle of the matter.

---

All that being said, if this truly doesn't currently exist in _any_ form - which I find hard to believe, I feel like I simply must not have found it yet? - then... I sadly don't have the resources necessary to incentivize or kickstart its development. I can't expect anybody to do that for me, because I'd be expecting them to do it for free.

I'm just thinking out loud, here... what about something like y.js? true P2P collaboration - even with advanced functionality - _without_ an intermediary server is _clearly_ an extant possibility. What I want to know is why has nobody seemingly implemented such a solution within a text editor or IDE yet?

Obviously there must be a good answer to _that_ question, but I can't think of one other than that it's simply not seen as a priority... which I find confusing but understandable - given that collaborative code is relatively new, I can understand why people haven't begun to see it as essential enough to develop a decentralized solution for... even though that disappoints me greatly. I'm just not sure where to go from that point, I suppose.

---

So.

Philosophizing aside, TL;DR: I desire a FOSS solution largely to save money; so that I can know it will actually be available to me in perpetuity, on premises, _not at somebody else's expense_; and so that can at least theoretically be adapted to my unusually picky needs and preferences if I were to either somehow learn how to do so myself or eventually afford the services of a developer. But of course if that truly, somehow, doesn't already exist... then it's not something I can reasonably ask of somebody to create.

I'd be highly disappointed and somewhat surprised if this kind of implementation doesn't already exist _somewhere_ in _some_ form, but if it doesn't... I guess I understand? Even though I don't quite get why I'd (in that case) be in the minority in placing it high in my list of priorities for a text editor.

**Comment by andrewsali** [comment]
JupyterLab has  [real time collaboration](https://jupyterlab.readthedocs.io/en/stable/user/rtc.html) using y.js, it would be nice if code server supported something similar.

**Comment by toddewilliams** [comment]
@morganwk Thank you very much! That was an expansive reply that really does show how conflicting some of these decisions and desires can be.

I've always wondered what the principle concerns are for people that hold the apparently conflicting opinions of "I want all the software I use to be FOSS" and "I'm a software developer and want to be paid for the software I write."

I'm at the phase of my software career where I have to make payroll for talented developers that remind me of my former self. So to pay them, and cover our other operating costs, we charge for some of our products. We believe it's an honest and direct financial model.

As to why there are no general OSS solutions: I'll postulate that it's because robust remote editing with multiple participants is extremely difficult to do in a correct and instantaneous manner. For a introduction of what's involved, this paper on [Differential Synchronization](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/35605.pdf) is a great primer on the topic. Then you have to build on this to do any number of participants, from any distance, with edit transmit times will vary wildly from edit to edit, who use different IDEs or browsers, etc.

But again, I really do appreciate you thinking through this so completely. I found your written internal dialog quite fascinating as it's from a completely different perspective than my own. 

I wish you the best in finding (or building) what you're looking for.

**Comment by stratosgear** [comment]
Apologies if I am oversimplifying things, but why would it be difficult to implement concurrent editing of a simple text document (or two, five, or twenty)?  Google Docs is doing it.  Sharelatex (now Overleaf) is doing it.  In the case of source code editing, it might even be more simplified since no document formatting is required (no headings, no images, no tables, no wysiwyg elements).  Actually the `y.js` library mentioned above seems to be solving all this (I was not aware of it, until 5 mins ago)

If the concurrent editing results in an invalid source file, let the compiler complain.  All the remaining infrastructure work that has already been implemented  (spinning servers up, having access to terminals, compiling, port forwards, fully onlne editors, etc) can remain the same.  I mean, the majority of, what I would consider the actually hard work, are already implemented and they do not have to support multiuser access.  Is it the simple editing of some text files that is really the roadblock here...?  (Note, I already apologized for potentially oversimplifying things :) )

Basically, I am also thinking out loudly about a similar internal lack of understanding of something like @morganwk  explained above...

**Comment by SichangHe** [comment]
This would be nice to be built into code-server itself, without the need for a separate hosting service.

**Comment by mictadlo** [comment]
Hi, I would like to be able to create two accounts one for my daughter and one for me. Would it be possible so we could  work/learn on our own projects with proper isolation?

**Comment by KaKi87** [comment]
You posted the same link twice 😅

**Comment by toddewilliams** [comment]
It's been awhile since I've tested it, but [CodeTogether Cloud](https://open-vsx.org/extension/genuitecllc/codetogether) works in VSCode and most derivatives. 

And it's free and works across different IDEs and even with a browser client.
